geomstats/geometry/siegel.py

```diff
@@ -0,0 +1,614 @@
+        kwargs.setdefault("metric", SiegelMetric(n))
+        super().__init__(
+            dim=n**2, embedding_space=ComplexMatrices(m=n, n=n), **kwargs
+        )
+        self.n = n
+        self.symmetric = symmetric
+        self.scale = scale
+
+    def belongs(self, point, atol=gs.atol):
+        """Check if a matrix belongs to the Siegel space.
+
+        Parameters
+        ----------
+        point : array-like, shape=[..., n, n]
+            Point to be checked.
+        atol : float
+            Tolerance.
+            Optional, default: backend atol.
+
+        Returns
+        -------
+        belongs : array-like, shape=[...,]
+            Boolean denoting if mat belongs to the Siegel space.
+        """
+        point_shape = gs.shape(point)
+
+        ndim = len(point_shape)
+        if ndim == 3:
+            n_samples = point_shape[0]
+
+        point_transconj = ComplexMatrices.transconjugate(point)
+        aux = gs.einsum("...ij,...jk->...ik", point, point_transconj)
+
+        if ndim == 2:
+            eigenvalues = gs.linalg.eigvalsh(aux)
+            belongs = gs.all(eigenvalues <= 1 - atol)
+        elif ndim == 3:
+            belongs = []
+            for i_sample in range(n_samples):
+                eigenvalues = gs.linalg.eigvalsh(aux[i_sample, ...])
+                belongs.append(gs.all(eigenvalues <= 1 - atol))
+            belongs = gs.array(belongs)
+
+        if self.symmetric:
+            belongs = gs.logical_and(belongs, ComplexMatrices.is_symmetric(point))
+
+        return belongs
+
+    def projection(self, point, atol=gs.atol):
+        """Project a matrix to the Siegel space.
+
+        Parameters
+        ----------
+        point : array-like, shape=[..., n, n]
+            Matrix to project.
+        atol : float
+            Tolerance.
+            Optional, default: backend atol.
+
+        Returns
+        -------
+        projected: array-like, shape=[..., n, n]
+            Matrix in the Siegel space.
+        """
+        data_type = point.dtype
+        point_shape = gs.shape(point)
+
+        ndim = len(point_shape)
+        if ndim == 3:
+            n_samples = point_shape[0]
+
+        if self.symmetric:
+            point = ComplexMatrices.to_symmetric(point)
+
+        point_transconj = ComplexMatrices.transconjugate(point)
+        aux = gs.einsum("...ij,...jk->...ik", point, point_transconj)
+
+        projected = point
+        eigenvalues = gs.linalg.eigvalsh(aux)
+        max_eigenvalues = gs.max(eigenvalues, axis=-1) ** 0.5
+
+        if ndim == 2:
+            if max_eigenvalues >= 1 - atol:
+                projected *= gs.cast((1 - atol) / max_eigenvalues, dtype=data_type)
+
+        elif ndim == 3:
+            for i_sample in range(n_samples):
+                if max_eigenvalues[i_sample] > 1 - atol:
+                    projected *= gs.cast(
+                        (1 - atol) / max_eigenvalues[i_sample], dtype=data_type
+                    )
```

Comment:
is this the behavior you really want? i.e. if max eigenvalue $i$ is greater than 1, then ALL the projected points will be multiplied by the scalar.

I think we can avoid the loop by using `gs.where`, where we set it to 1. in the places the condition is not verified.

Reply:
Thank you, I have replaced `projected` by `projected[i_sample]`.
I do not think that we can use `gs.where` here as the condition `max_eigenvalues[i_sample] > 1 - atol` does not have the same shape than the array `projected` which has the desired output shape.
Done.